riscv-asm.md

```diff
@@ -214,7 +214,150 @@ ATTRIBUTE_VALUE := <string>
+                        | <FULLARCHSTR>
+
+EXTENSIONS             := <EXTENSION> ',' <EXTENSIONS>
+                        | <EXTENSION>
+
+FULLARCHSTR            := <full-arch-string>
+
+EXTENSION              := <OP> <EXTENSION-NAME> [<VERSION>]
+
+OP                     := '+'
+                        | '-'
+
+VERSION                := [0-9]+ 'p' [0-9]+
+                        | [1-9][0-9]*
+
+EXTENSION-NAME         := Naming rule is defined in RISC-V ISA manual
+```
+
+- Extension version can be omitted, the assembler will use the built-in default
+  version for that extension.
+- `OP` can be enable (`+`) or disable (`-`).
+- Input format of `<full-arch-string>` is the same as `-march` option.
+
+Example:
+
+```assembly
+.attribute arch, "rv64imafdc"
+# You can only use instruction from i, m, a, f, d and c extensions.
+memcpy_general:
+    add     a5,a1,a2
+    beq     a1,a5,.L2
+    add     a2,a0,a2
+    mv      a5,a0
+.L3:
+    addi    a1,a1,1
+    addi    a5,a5,1
+    lbu     a4,-1(a1)
+    sb      a4,-1(a5)
+    bne     a5,a2,.L3
+.L2:
+    ret
+
+.option push     # Push current option to stack.
+.option arch, +v # Enable vector extension, we can use any instruction in imafdcv extension.
+memcpy_vec:
+    mv a3, a0
+.Lloop:
+    vsetvli t0, a2, e8, m8, ta, ma
+    vle8.v v0, (a1)
+    add a1, a1, t0
+    sub a2, a2, t0
+    vse8.v v0, (a3)
+    add a3, a3, t0
+    bnez a2, .Lloop
+    ret
+.option pop   # Pop current option from stack, restore the enabled ISA extension status to imafdc.
+
+.option push     # Push current option to stack.
+.option arch, -c # Disable compressed extension, we can't use any instruction in extension.
+memcpy_norvc:
+    add     a5,a1,a2
+    beq     a1,a5,.L2
+    add     a2,a0,a2
+    mv      a5,a0
+.L3:
+    addi    a1,a1,1
+    addi    a5,a5,1
+    lbu     a4,-1(a1)
+    sb      a4,-1(a5)
+    bne     a5,a2,.L3
+.L2:
+    ret
+.option pop   # Pop current option from stack, restore the enabled ISA extension status to imafdc.
+
+.option push           # Push current option to stack.
+.option arch, rv64imc # Set arch to rv64imc.
+    nop
+.option pop   # Pop current option from stack, restore the enabled ISA extension status to imafdc.
+
+```
+
+NOTE: A typical use case is with `ifunc`, e.g. the C library is built with
+`rv64gc`, but a few functions like memcpy provide two versions, one built with
+`rv64gc` and one built with `rv64gcv`, and then select between them by ifunc
+mechanism at run-time.  However, we don't want to change the minimal execution
+environment requirement to `rv64gcv`, since the `rv64gcv` version will be
+invoked only if the execution environment supports the vector extension, so
+the minimal execution environment requirement still is `rv64gc`.
+
+NOTE: `.option arch, +` will also enable all required extensions, for example,
+`rv32i` + `.option arch, +v` will also enable `f`, `d`, `zve32x`, `zve32f`,
+`zve64x`, `zve64f`, `zve64d`, `zvl32b`, `zvl64b` and `zvl128b` extensions.
```

Comment:
It would be good to explain somewhere that e.g. `+c, -c` and `+v, -v` do the expected things. I guess the expected thing is that the named extension is disabled but its dependencies are enabled.

Reply:
(alternatively it could be an error)